### DISCUSSION TITLE
metrics loading failed with error "undefined"

### DISCUSSION THREAD

**Discussion by ljharb** [lowlighter/metrics#996] [Q&A]
https://github.com/ljharb/ljharb/runs/6066500430?check_suite_focus=true

any ideas?

**Comment by lowlighter** [comment]
Hello 👋 !

The action is trying to perform a `HEAD` request on `https://api.github.com` to validate that provided token is correct and is able to fetch data from it, though it failed because it seems you reached the GitHub rate limit: (`x-ratelimit-remaining: 0`, with `5013` requests consumed, resets at `1650306689`):

```js
response: {
  url: 'https://api.github.com/',
  status: 403,
  headers: {
    'x-github-request-id': '0400:3D5E:A5E71:B5198:625DAC7C',
    'x-ratelimit-limit': '5000',
    'x-ratelimit-remaining': '0',
    'x-ratelimit-reset': '1650306689',
    'x-ratelimit-used': '5013',
  },
  data: undefined
},
request: {
  method: 'HEAD',
  url: 'https://api.github.com/',
  headers: {
    accept: 'application/vnd.github.v3+json',
    'user-agent': 'octokit-core.js/3.5.1 Node.js/17.3.1 (linux; x64)',
    authorization: 'token [REDACTED]'
  },
}
```

If you use your account for other application that use GitHub APIs, I think the solution would be to create a bot account and use its personal access token instead (you can change the target user by providing the [`user`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#user) option)

For info here's what your current config will look like (I generated it with my own token):

<img src="https://gist.githubusercontent.com/lowlighter/6cba4b038ec95cdfcdbaca2bdbe1c438/raw/7abb9ed1657224e634feba18bcaf3518895940f1/github-metrics.svg">

Also JFY the `license` plugin is intended for repositories (it checks license of dependencies, like those provided in a `package.json`) so it won't render when used on a user profile 🙂 

Hope this helps, and thanks a lot for giving this action a try 😄 !

**Reply by ljharb** [reply]
ah ok, thanks. i'll remove that plugin and tweak the config based on that output.

i definitely use it on a ton of other things, but a bot account wouldn't have the same access my user account would, unfortunately, so I might just have to make it run less often and ignore ratelimit failures.

Could there perhaps be a) a better error message in that case, and b) a config that lets the action *succeed* when that error happens? That way I could run it on a cron, and ignore all the rate limit failures, and trust that it would "eventually" work.

**Reply by ljharb** [reply]
(also, while i have your attention, i'd love a way to include my actual full streak info, not just a year's worth - that seems like something a plugin could do?)

**Reply by lowlighter** [reply]
> a bot account wouldn't have the same access my user account would

While it's true, a lot of data can still be extracted from an external account (as you can see from the above output), especially if you don't care about private repositories stats. (though the sponsors plugin does requires your own token for some data, since it's exposed only for your account by GitHub for obvious privacy reason)

I noticed you've enabled the `plugin_sponsors_past` option, which is intended to display expired/past sponsors, but currently this option is inseparable from the "sponsorship goal card" so it won't have any effects on the current version. It's also limited to the past 100 sponsorships events.

The sponsors plugin is probably not the most used one so I didn't get a lot of feedback for it (so it's mostly tailored for my needs), but usually I'm open to feature requests and ideas suggestions if they're feasible 👍 

> Could there perhaps be a) a better error message in that case, and b) a config that lets the action succeed when that error happens? 

Yeah indeed, this error is kind of hard to debug, I'll check to better handle it 👍 

___

Also on a totatlly unrelated side note, thanks a lot for what you're doing for EcmaScript. I really admire your dedication 😄

**Reply by ljharb** [reply]
For the sponsors plugin, tbh i'd love just a facepile of current sponsors, and another facepile of past sponsors (obv pursuant to the individual's sponsorship being public).

**Reply by lowlighter** [reply]
> (also, while i have your attention, i'd love a way to include my actual full streak info, not just a year's worth - that seems like something a plugin could do?)

Will be tracked in #1002, but maybe I'll add a new "regular calendar" plugin since it may look really cluttered with the isometric view

> For the sponsors plugin, tbh i'd love just a facepile of current sponsors, and another facepile of past sponsors (obv pursuant to the individual's sponsorship being public).

Will be tracked in #1000

Thanks a lot for your feedback and suggestions 👍
